### PR TITLE
Compute min length mempool transactions message by type

### LINKED ISSUE #6852: Opt-out early of decoding on too short input
### Describe the feature

Opting out of decoding transactions can be done earlier than it is done, based on the minimum possible length of an encoded transaction.

https://github.com/paradigmxyz/reth/blob/d36180deca1ac4bfd4ec943d512aff554d4f7a98/crates/primitives/src/transaction/mod.rs#L1385

The minimum length of an encoded transaction is type dependent. This code is executed very often hence the optimisation is valuable.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by emhane**
Closes https://github.com/paradigmxyz/reth/issues/6852.

Calculates the minimum possible length of the rlp encoded transaction types by tests. Uses these values to opt-out early of decoding.

### CODE REVIEW

File: crates/primitives/src/transaction/mod.rs [hunk 1]
```diff
@@ -71,6 +71,18 @@ pub(crate) static PARALLEL_SENDER_RECOVERY_THRESHOLD: Lazy<usize> =
...
+pub const MIN_LENGTH_LEGACY_TX_ENCODED: usize = 10;
+/// Minimum length of a rlp-encoded eip2930 transaction.
+pub const MIN_LENGTH_EIP2930_TX_ENCODED: usize = 14;
+/// Minimum length of a rlp-encoded eip1559 transaction.
+pub const MIN_LENGTH_EIP1559_TX_ENCODED: usize = 15;
+/// Minimum length of a rlp-encoded eip4844 transaction.
+pub const MIN_LENGTH_EIP4844_TX_ENCODED: usize = 17;
+/// Minimum length of a rlp-encoded deposit transaction.
+#[cfg(feature = "optimism")]
+pub const MIN_LENGTH_DEPOSIT_TX_ENCODED: usize = 65;
```
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:84]: how did we get these?

**Review by mattsse — COMMENTED** [review]
I think that is reasonable, though perhaps unnecessary because decoding would fail regardless.

This adds a (tiny) overhead to decoding and not much benefit, so I'm a bit on the fence.

wdyt @Rjected

File: crates/primitives/src/transaction/mod.rs (hunk 1 above)
Comment (emhane) [reply crates/primitives/src/transaction/mod.rs:84]: the tests as mentioned in the description of this pr

**Review by Rjected — COMMENTED** [review]
yeah, while this causes invalid transactions to fail earlier, it adds overhead for decoding any valid transaction. By the second field most short transaction types already fail, and by the third, everything but deposit would fail. And it isn't slow to decode these fields. So I'm not sure that the benefits outweigh the costs here

**Comment by emhane** [comment]
> yeah, while this causes invalid transactions to fail earlier, it adds overhead for decoding any valid transaction. By the second field most short transaction types already fail, and by the third, everything but deposit would fail. And it isn't slow to decode these fields. So I'm not sure that the benefits outweigh the costs here

hm, it's just reading one byte though to get the type...

**Comment by emhane** [comment]
alright will remove from decoding and keep constants + tests for tx fetcher validation (incorporation in validation will be other pr).

File: crates/primitives/src/transaction/mod.rs [hunk 2]
```diff
@@ -1251,20 +1264,23 @@ impl TransactionSigned {
...
-            2 => Transaction::Eip1559(TxEip1559::decode_inner(data)?),
-            3 => Transaction::Eip4844(TxEip4844::decode_inner(data)?),
+            TxType::EIP2930 => Transaction::Eip2930(TxEip2930::decode_inner(data)?),
+            TxType::EIP1559 => Transaction::Eip1559(TxEip1559::decode_inner(data)?),
+            TxType::EIP4844 => Transaction::Eip4844(TxEip4844::decode_inner(data)?),
             #[cfg(feature = "optimism")]
-            0x7E => Transaction::Deposit(TxDeposit::decode_inner(data)?),
-            _ => return Err(RlpError::Custom("unsupported typed transaction type")),
+            TxType::DEPOSIT => Transaction::Deposit(TxDeposit::decode_inner(data)?),
+            TxType::Legacy => unreachable!("path for legacy tx has diverged before this method"),
```
Comment (Rjected) [inline crates/primitives/src/transaction/mod.rs:1276]: this message is better yeah

File: crates/primitives/src/transaction/mod.rs [hunk 3]
```diff
@@ -1928,4 +1947,106 @@ mod tests {
...
+
+        let signed_tx = TransactionSigned::from_transaction_and_signature(
+            Transaction::Legacy(transaction),
+            signature,
+        );
+
+        let mut encoded = BytesMut::new();
+        signed_tx.encode(&mut encoded);
+
+        assert_eq!(b"\xc9\x80\x80\x80\x80\x80\x80\x1b\x80\x80", &encoded[..]);
```
Comment (Rjected) [inline crates/primitives/src/transaction/mod.rs:1964]: we also have the `hex!` macro in alloy if it makes this easier

**Review by Rjected — APPROVED** [review]
lgtm!

**Pull request merged**